presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
         fileIterators.addLast(iterator);
     }
 
-    private static List<LocatedFileStatus> listAndSortBucketFiles(HiveFileIterator hiveFileIterator, int bucketCount)
+    private List<Set<LocatedFileStatus>> getBucketFileSets(HiveFileIterator hiveFileIterator, int bucketCount)
     {
-        ArrayList<LocatedFileStatus> list = new ArrayList<>(bucketCount);
-
+        List<LocatedFileStatus> fileStatuses = new ArrayList<>(bucketCount);
         while (hiveFileIterator.hasNext()) {
             LocatedFileStatus next = hiveFileIterator.next();
             if (isDirectory(next)) {
                 // Fail here to be on the safe side. This seems to be the same as what Hive does
                 throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("%s Found sub-directory in bucket directory for partition: %s", CORRUPT_BUCKETING, hiveFileIterator.getPartitionName()));
             }
-            list.add(next);
+            fileStatuses.add(next);
         }
 
-        if (list.size() != bucketCount) {
-            throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("%s The number of files in the directory (%s) does not match the declared bucket count (%s) for partition: %s", CORRUPT_BUCKETING, list.size(), bucketCount, hiveFileIterator.getPartitionName()));
+        // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
+        fileStatuses.sort(null);
+
+        List<Set<LocatedFileStatus>> list = null;
+        if (fileStatuses.size() == bucketCount) {
+            list = getBucketFilesSetsSingle(fileStatuses);
+        }
+        else if (isMultiFileBucketingEnabled(session)) {
+            list = getBucketFilesSetsMulti(fileStatuses);
+            if (list.size() != bucketCount) {
+                list = null;
```

Comment:
I don't like idea of using list as control variable.
I would either use well named bool for that or just simply assign value to a list like:
```
if (isMultiFileBucketingEnabled(session) {
  list = getBucketFilesSetsMulti(fileStatuses);
} 
else {
  list = getBucketFilesSetsSingle(fileStatuses);
}
```

and then 
```
if (list.size() != bucketCount) {
  throw new PrestoException(HIVE_INVALID_BUCKET_FILES, "OMG, OMG, OMG, Something went wrong");
}
```